store/sqldb/traffic.go

```diff
@@ -55,108 +56,164 @@ func (s *SpDBImpl) CheckQuotaAndAddReadRecord(record *corespdb.ReadRecord, quota
+}
 
-	// add read record
-	insertReadRecord := &ReadRecordTable{
-		BucketID:        record.BucketID,
-		ObjectID:        record.ObjectID,
-		UserAddress:     record.UserAddress,
-		ReadTimestampUs: record.ReadTimestampUs,
-		BucketName:      record.BucketName,
-		ObjectName:      record.ObjectName,
-		ReadSize:        record.ReadSize,
-	}
-	result = s.db.Create(insertReadRecord)
-	if result.Error != nil || result.RowsAffected != 1 {
-		err = fmt.Errorf("failed to insert read record table: %s", result.Error)
-		return err
-	}
-	return nil
+// updateConsumedQuota update the consumed quota of BucketTraffic table in the transaction way
+func (s *SpDBImpl) updateConsumedQuota(record *corespdb.ReadRecord, quota *corespdb.BucketQuota) error {
+	err := s.db.Transaction(func(tx *gorm.DB) error {
+		var bucketTraffic BucketTrafficTable
+		var err error
+		if err = tx.Clauses(clause.Locking{Strength: "UPDATE"}).Where("bucket_id = ?", record.BucketID).Find(&bucketTraffic).Error; err != nil {
+			return fmt.Errorf("failed to query bucket traffic table: %v", err)
+		}
+
+		// if charged quota changed, update the new value
+		if bucketTraffic.ChargedQuotaSize != quota.ChargedQuotaSize {
+			result := tx.Model(&bucketTraffic).
+				Updates(BucketTrafficTable{
+					ChargedQuotaSize: quota.ChargedQuotaSize,
+					ModifiedTime:     time.Now(),
+				})
+			if result.Error != nil {
+				return fmt.Errorf("failed to update bucket traffic table: %s", result.Error)
+			}
+
+			if result.RowsAffected != 1 {
+				return fmt.Errorf("update traffic of %s has affected more than one rows %d, "+
+					"update charged quota %d", bucketTraffic.BucketName, result.RowsAffected, quota.ChargedQuotaSize)
+			}
+		}
+
+		updatedReadConsumedSize, updatedFreeConsumedSize, err := getUpdatedConsumedQuota(record, bucketTraffic.FreeQuotaSize, bucketTraffic.FreeQuotaConsumedSize,
+			bucketTraffic.ReadConsumedSize, bucketTraffic.ChargedQuotaSize)
+		if err != nil {
+			return err
+		}
+
+		if err = tx.Model(&bucketTraffic).
+			Updates(BucketTrafficTable{
+				ReadConsumedSize:      updatedReadConsumedSize,
+				FreeQuotaConsumedSize: updatedFreeConsumedSize,
+				ModifiedTime:          time.Now(),
+			}).Error; err != nil {
+			return fmt.Errorf("failed to update bucket traffic table: %v", err)
+		}
+		// commit transaction
+		if err = tx.Commit().Error; err != nil {
+			return fmt.Errorf("failed to commit bucket traffic table: %v", err)
+		}
+
+		return nil
+	})
+
+	return err
 }
 
+// InitBucketTraffic init the bucket traffic table
 func (s *SpDBImpl) InitBucketTraffic(bucketID uint64, bucketName string, quota *corespdb.BucketQuota) error {
-	insertBucketTraffic := &BucketTrafficTable{
-		BucketID:              bucketID,
-		FreeQuotaSize:         quota.FreeQuotaSize,
-		FreeQuotaConsumedSize: 0,
-		BucketName:            bucketName,
-		ReadConsumedSize:      0,
-		ChargedQuotaSize:      quota.ChargedQuotaSize,
-		ModifiedTime:          time.Now(),
+	tx := s.db.Begin()
```

Comment:
why not use the same code style `s.db.Transaction(func(tx *gorm.DB) error {...}`

Reply:
updated